oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/IndexUpdate.java

```diff
@@ -330,9 +334,14 @@ private void collectIndexEditors(NodeBuilder definitions,
                     // then we don't need to handle missing handler
                     if (definition.hasProperty(ASYNC_PROPERTY_NAME) && rootState.async == null) {
                         if (!TYPE_DISABLED.equals(type)) {
-                            log.warn("Missing provider for nrt/sync index: {} (rootState.async: {}). " +
-                                    "Please note, it means that index data should be trusted only after this index " +
-                                    "is processed in an async indexing cycle.", definition, rootState.async);
+                            long now = System.currentTimeMillis();
+                            if (now > lastMissingProviderMessageTime + 60 * 1000) {
+                                lastMissingProviderMessageTime = now;
```

Comment:
The read-test-update of `lastMissingProviderMessageTime` is not atomic, so two or more threads may pass the check and print the log more or less at the same time. If this is an intentional loosening of the specification of "log at most every minute" for the sake of performance, please write it in the comments. Otherwise, I think it is more clear and clean if the implementation did not have race conditions that allow it to break the intention that is apparent from the code (log at most once a minute).

Reply:
It isn't for performance, but for simplicity. In my view, race conditions in this area (logging) are not a problem: worst case, we would log multiple times; that is fine. Sure, if there is a way to do this without adding complexity that would be nice as well! I'll try with AtomicLong.

Reply:
Ok, I also can't think of a way that ensures that we log only once per second using atomic variable. But could you just add a comment to say that the code does not ensure that we log at most once per second but that is ok in this case? So that our future selves or other people will not look at the code and be misled by it.

Reply:
I think this would work and is not too complicated. 

```java
private static volatile AtomicLong lastMissingProviderMessageTime = new AtomicLong(0);


  long now = System.currentTimeMillis();
  long previous = lastMissingProviderMessageTime.get();
  if (now > previous + 60 * 1000) {
      if (lastMissingProviderMessageTime.compareAndSet(previous, now)) {
         // log
      }
```
But leaving it here mainly for intelectual curiosity, could be useful in other situations.